sretoolbox/binaries/base.py

```diff
@@ -0,0 +1,243 @@
+"""
+Abstractions around system binaries.
+"""
+
+import logging
+import os
+
+from abc import ABCMeta
+from abc import abstractmethod
+from collections import Counter
+from pathlib import Path
+
+import requests
+
+from semver import VersionInfo
+
+from sretoolbox.utils import run
+
+
+LOG = logging.getLogger(__name__)
+
+
+class CommandNotFoundError(Exception):
+    """
+    Used when a binary is not available in the system.
+    """
+
+
+class Binary(metaclass=ABCMeta):
+    """
+    Represents a binary in the system. Tries to find the system binary
+    on the $PATH, in the specified version. When the binary is not
+    available locally, tryes to download it.
+
+    :param version: the version of the binary
+    """
+    binary_template = ''
+    download_url_template = ''
+
+    def __init__(self, version, download_path='/tmp'):
+        # Making sure that the version contains
+        # valid minor and patch elements
+        counter = Counter(version)
+        if counter['.'] == 0:
+            version += '.0.0'
+        elif counter['.'] == 1:
+            version += '.0'
+
```

Comment:
I guess this logic will work if the version is an integer... Can we expect something different?

Reply:
version is a string. I will make that clear in the docstring, thanks for pointing out.

Reply:
I mean, what happens if version is something like `1.14-pre1`?